jlm/rvsdg/structural-node.hpp

```diff
@@ -77,20 +77,29 @@ class structural_input : public node_input
 public:
   virtual ~structural_input() noexcept;
 
-protected:
```

Comment:
Why did you remove it?

Reply:
make_shared (and make_unique) need a public constructor, otherwise they cannot construct the class. It is also essential to use make_shared<X>(...) instead of shared_ptr<X>(new X(...)) as the former will cause one allocation, the latter causes two

if the constructor needs to be made "really" private (but you generally do not need that, unless you have something like enabled_shared_from_this), it requires some construction like:

`class X {
private:
  struct private_tag {};
public:
  X(Arg1 arg1, Arg2 arg2, ..., private_tag);

  shared_ptr<X> Create(Arg1 arg1, Arg2 arg2, ...) { return make_shared<X>(arg1, arg2, ..., private_tag {}); }
`

but that is a bit much for now, also since I need to do some more cleanup later